File: frontend/static/js/bookmarks.js
// Bookmarks management functionality

// Load and display bookmarks
async function loadBookmarks() {
    try {
        const response = await axios.get('/api/bookmarks/');
        const groups = response.data;

        const container = document.getElementById('bookmarksContainer');
        container.innerHTML = '';

        groups.forEach(group => {
            const groupCard = createBookmarkGroupCard(group);
            container.appendChild(groupCard);
        });

        // Setup drag and drop for bookmarks
        setupBookmarksDragDrop();
    } catch (error) {
        console.error('Error loading bookmarks:', error);
        showToast('Failed to load bookmarks', 'error');
    }
}

// Create bookmark group card
function createBookmarkGroupCard(group) {
    const card = document.createElement('div');
    card.className = 'card bookmark-group-card mb-3';
    card.dataset.group = group.name;

    // Escape single quotes in group name for onclick handlers
    const escapedGroupName = group.name.replace(/'/g, "\\'");

    card.innerHTML = `
        <div class="bookmark-group-header">
            <span class="drag-handle">⋮</span>
            <span class="group-title">${group.name}</span>
            <div class="group-actions">
                <button class="btn btn-sm btn-primary" onclick="showAddBookmarkModal('${escapedGroupName}')">
                    <i class="bi bi-plus"></i>
                </button>
                <button class="btn btn-sm btn-outline-primary" onclick="renameBookmarkGroup('${escapedGroupName}')">
                    <i class="bi bi-pencil"></i>
                </button>
                <button class="btn btn-sm btn-danger" onclick="deleteBookmarkGroup('${escapedGroupName}')">
                    <i class="bi bi-trash"></i>
                </button>
            </div>
        </div>
        <div class="bookmark-group-body" id="group-${group.name}">
            ${group.bookmarks.map(bookmark => createBookmarkItem(group.name, bookmark)).join('')}
        </div>
    `;

    return card;
}

// Create individual bookmark item
function createBookmarkItem(groupName, bookmark) {
    // Escape single quotes for onclick handlers
    const escapedGroupName = groupName.replace(/'/g, "\\'");
    const escapedBookmarkName = bookmark.name.replace(/'/g, "\\'");

    return `
        <div class="bookmark-item" data-bookmark="${bookmark.name}">
            <span class="bookmark-drag-handle">⋮</span>
            ${bookmark.icon ? `<img src="${bookmark.icon}" class="bookmark-icon" alt="">` : ''}
            <div class="bookmark-details">
                <a href="${bookmark.href}" target="_blank" class="bookmark-name">${bookmark.name}</a>
                ${bookmark.description ? `<small class="text-muted">${bookmark.description}</small>` : ''}
            </div>
            <div class="bookmark-actions">
                <button class="btn btn-sm btn-link" onclick="editBookmark('${escapedGroupName}', '${escapedBookmarkName}')">
                    <i class="bi bi-pencil"></i>
                </button>
                <button class="btn btn-sm btn-link text-danger" onclick="deleteBookmark('${escapedGroupName}', '${escapedBookmarkName}')">
                    <i class="bi bi-trash"></i>
                </button>
            </div>
        </div>
    `;
}

// Show add bookmark modal
function showAddBookmarkModal(groupName) {
    const modal = new bootstrap.Modal(document.getElementById('bookmarkModal'));
    document.getElementById('bookmarkModalTitle').textContent = 'Add Bookmark';
    document.getElementById('bookmarkForm').reset();
    document.getElementById('bookmarkGroup').value = groupName;
    document.getElementById('bookmarkModalAction').textContent = 'Add';
    document.getElementById('bookmarkForm').dataset.mode = 'add';
    modal.show();
}

// Edit bookmark
async function editBookmark(groupName, bookmarkName) {
    try {
        const encodedGroupName = encodeURIComponent(groupName);
        const response = await axios.get(`/api/bookmarks/${encodedGroupName}`);
        const bookmark = response.data.bookmarks.find(b => b.name === bookmarkName);

        if (!bookmark) {
            showToast('Bookmark not found', 'error');
            return;
        }

        const modal = new bootstrap.Modal(document.getElementById('bookmarkModal'));
        document.getElementById('bookmarkModalTitle').textContent = 'Edit Bookmark';
        document.getElementById('bookmarkGroup').value = groupName;
        document.getElementById('bookmarkName').value = bookmark.name;
        document.getElementById('bookmarkHref').value = bookmark.href;
        document.getElementById('bookmarkIcon').value = bookmark.icon || '';
        document.getElementById('bookmarkDescription').value = bookmark.description || '';
        document.getElementById('bookmarkModalAction').textContent = 'Update';
        document.getElementById('bookmarkForm').dataset.mode = 'edit';
        document.getElementById('bookmarkForm').dataset.originalName = bookmarkName;
        modal.show();
    } catch (error) {
        console.error('Error loading bookmark:', error);
        showToast('Failed to load bookmark', 'error');
    }
}

// Save bookmark (add or update)
async function saveBookmark() {
    const form = document.getElementById('bookmarkForm');
    const mode = form.dataset.mode;
    const groupName = document.getElementById('bookmarkGroup').value;
    const bookmarkData = {
        name: document.getElementById('bookmarkName').value,
        href: document.getElementById('bookmarkHref').value,
        icon: document.getElementById('bookmarkIcon').value || null,
        description: document.getElementById('bookmarkDescription').value || null
    };

    try {
        const encodedGroupName = encodeURIComponent(groupName);
        if (mode === 'add') {
            await axios.post(`/api/bookmarks/${encodedGroupName}`, bookmarkData);
        } else {
            const originalName = form.dataset.originalName;
            const encodedOriginalName = encodeURIComponent(originalName);
            await axios.put(`/api/bookmarks/${encodedGroupName}/${encodedOriginalName}`, bookmarkData);
        }

        showToast(`Bookmark ${mode === 'add' ? 'added' : 'updated'} successfully`, 'success');
        bootstrap.Modal.getInstance(document.getElementById('bookmarkModal')).hide();
        loadBookmarks();
    } catch (error) {
        console.error('Error saving bookmark:', error);
        showToast('Failed to save bookmark', 'error');
    }
}

// Delete bookmark
async function deleteBookmark(groupName, bookmarkName) {
    if (!confirm(`Are you sure you want to delete "${bookmarkName}"?`)) {
        return;
    }

    try {
        const encodedGroupName = encodeURIComponent(groupName);
        const encodedBookmarkName = encodeURIComponent(bookmarkName);
        await axios.delete(`/api/bookmarks/${encodedGroupName}/${encodedBookmarkName}`);
        showToast('Bookmark deleted successfully', 'success');
        loadBookmarks();
    } catch (error) {
        console.error('Error deleting bookmark:', error);
        showToast('Failed to delete bookmark', 'error');
    }
}

// Create new bookmark group
async function createBookmarkGroup() {
    const groupName = prompt('Enter group name:');
    if (!groupName) return;

    try {
        const encodedGroupName = encodeURIComponent(groupName);
        await axios.post(`/api/bookmarks/groups/${encodedGroupName}`);
        showToast('Group created successfully', 'success');
        loadBookmarks();
    } catch (error) {
        console.error('Error creating group:', error);
        showToast(error.response?.data?.detail || 'Failed to create group', 'error');
    }
}

// Rename bookmark group
async function renameBookmarkGroup(oldName) {
    console.log('Renaming group:', oldName);
    const newName = prompt(`Rename group "${oldName}" to:`, oldName);
    if (!newName || newName === oldName) return;

    try {
        // Encode the group name for URL
        const encodedOldName = encodeURIComponent(oldName);
        console.log('Encoded old name:', encodedOldName);
        await axios.put(`/api/bookmarks/groups/${encodedOldName}`, { new_name: newName });
        showToast('Group renamed successfully', 'success');
        loadBookmarks();
    } catch (error) {
        console.error('Error renaming group:', error);
        showToast(error.response?.data?.detail || 'Failed to rename group', 'error');
    }
}

// Delete bookmark group
async function deleteBookmarkGroup(groupName) {
    if (!confirm(`Are you sure you want to delete the group "${groupName}" and all its bookmarks?`)) {
        return;
    }

    try {
        // Encode the group name for URL
        const encodedGroupName = encodeURIComponent(groupName);
        await axios.delete(`/api/bookmarks/groups/${encodedGroupName}`);
        showToast('Group deleted successfully', 'success');
        loadBookmarks();
    } catch (error) {
        console.error('Error deleting group:', error);
        showToast(error.response?.data?.detail || 'Failed to delete group', 'error');
    }
}

// Export bookmarks
async function exportBookmarks() {
    try {
        const response = await axios.get('/api/bookmarks/export', {
            responseType: 'blob'
        });
        const blob = new Blob([response.data], { type: 'application/x-yaml' });
        const url = window.URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'bookmarks.yaml';
        document.body.appendChild(a);
        a.click();
        window.URL.revokeObjectURL(url);
        document.body.removeChild(a);
        showToast('Bookmarks exported successfully', 'success');
    } catch (error) {
        console.error('Error exporting bookmarks:', error);
        showToast('Failed to export bookmarks', 'error');
    }
}

// Import bookmarks
async function importBookmarks() {
    const input = document.createElement('input');
    input.type = 'file';
    input.accept = '.yaml,.yml';

    input.onchange = async (event) => {
        const file = event.target.files[0];
        if (!file) return;

        const formData = new FormData();
        formData.append('file', file);

        try {
            const response = await axios.post('/api/bookmarks/import', formData, {
                headers: { 'Content-Type': 'multipart/form-data' }
            });

            showToast(`Imported ${response.data.groups} groups with ${response.data.total_bookmarks} bookmarks`, 'success');
            loadBookmarks();
        } catch (error) {
            console.error('Error importing bookmarks:', error);
            showToast(error.response?.data?.detail || 'Failed to import bookmarks', 'error');
        }
    };

    input.click();
}

// Setup drag and drop for bookmarks
function setupBookmarksDragDrop() {
    // Implement drag and drop functionality for bookmark groups and items
    const groups = document.querySelectorAll('.bookmark-group-card');
    const groupContainer = document.getElementById('bookmarksContainer');

    if (groupContainer && groups.length > 0) {
        new Sortable(groupContainer, {
            animation: 150,
            handle: '.drag-handle',
            ghostClass: 'dragging',
            onEnd: function(evt) {
                // Save new order
                saveBookmarkOrder();
            }
        });

        // Setup sortable for bookmark items within groups
        groups.forEach(group => {
            const groupName = group.dataset.group;
            const itemsContainer = document.getElementById(`group-${groupName}`);

            if (itemsContainer) {
                new Sortable(itemsContainer, {
                    animation: 150,
                    handle: '.bookmark-drag-handle',
                    group: 'bookmarks',
                    ghostClass: 'dragging',
                    onEnd: function(evt) {
                        // Save new order
                        saveBookmarkOrder();
                    }
                });
            }
        });
    }
}

// Save bookmark order (implement if needed)
async function saveBookmarkOrder() {
    // This would save the new order to the backend
    console.log('Bookmark order changed');
}

// Show toast notification
function showToast(message, type = 'info') {
    // Implementation depends on your toast library
    console.log(`${type}: ${message}`);
    // You can use Bootstrap toasts or another notification library
}

File: backend/api/bookmarks.py
from fastapi import APIRouter, HTTPException, Body, UploadFile, File
from fastapi.responses import Response
from typing import List, Dict, Any
from models import BookmarkCreate, BookmarkUpdate
from core.bookmarks_handler import BookmarksHandler

router = APIRouter()
bookmarks_handler = BookmarksHandler()

@router.get("/groups")
async def get_bookmark_groups():
    """Get all bookmark groups"""
    groups = bookmarks_handler.get_all_groups()
    return {"groups": groups}

@router.get("/")
async def get_all_bookmarks():
    """Get all bookmarks organized by groups"""
    bookmarks = bookmarks_handler.load_bookmarks()
    groups = bookmarks_handler.parse_bookmarks(bookmarks)

    # Format response
    response = []
    for group_name, bookmarks_list in groups.items():
        group_data = {
            "name": group_name,
            "bookmarks": []
        }
        for bookmark in bookmarks_list:
            bookmark_data = {
                "name": bookmark['name'],
                **bookmark.get('config', {})
            }
            group_data["bookmarks"].append(bookmark_data)
        response.append(group_data)

    return response

@router.get("/{group}")
async def get_group_bookmarks(group: str):
    """Get bookmarks for a specific group"""
    bookmarks = bookmarks_handler.load_bookmarks()
    groups = bookmarks_handler.parse_bookmarks(bookmarks)

    if group not in groups:
        raise HTTPException(status_code=404, detail=f"Group '{group}' not found")

    bookmarks_list = []
    for bookmark in groups[group]:
        bookmark_data = {
            "name": bookmark['name'],
            **bookmark.get('config', {})
        }
        bookmarks_list.append(bookmark_data)

    return {
        "group": group,
        "bookmarks": bookmarks_list
    }

@router.post("/{group}")
async def create_bookmark(group: str, bookmark: BookmarkCreate):
    """Add a bookmark to a group"""
    bookmark_config = {
        "abbr": bookmark.name,  # Add abbr field to match Homepage format
        "href": bookmark.href
    }

    if bookmark.icon:
        bookmark_config["icon"] = bookmark.icon
    if bookmark.description:
        bookmark_config["description"] = bookmark.description

    success = bookmarks_handler.add_bookmark(group, bookmark.name, bookmark_config)

    if not success:
        raise HTTPException(status_code=400, detail="Bookmark already exists or could not be added")

    return {"message": "Bookmark created successfully"}

@router.put("/{group}/{bookmark_name}")
async def update_bookmark(group: str, bookmark_name: str, bookmark: BookmarkUpdate):
    """Update a bookmark"""
    bookmarks = bookmarks_handler.load_bookmarks()
    groups = bookmarks_handler.parse_bookmarks(bookmarks)

    if group not in groups:
        raise HTTPException(status_code=404, detail=f"Group '{group}' not found")

    # Find existing bookmark
    existing = None
    for b in groups[group]:
        if b['name'] == bookmark_name:
            existing = b
            break

    if not existing:
        raise HTTPException(status_code=404, detail=f"Bookmark '{bookmark_name}' not found")

    # Update config
    bookmark_config = existing.get('config', {})

    # Update name if changed
    if bookmark.name and bookmark.name != bookmark_name:
        bookmark_config["abbr"] = bookmark.name
    elif "abbr" not in bookmark_config:
        bookmark_config["abbr"] = bookmark_name

    if bookmark.href is not None:
        bookmark_config["href"] = bookmark.href
    if bookmark.icon is not None:
        bookmark_config["icon"] = bookmark.icon
    if bookmark.description is not None:
        bookmark_config["description"] = bookmark.description

    # If name is being changed
    new_name = bookmark.name if bookmark.name else bookmark_name

    if new_name != bookmark_name:
        # Delete old and add new
        bookmarks_handler.delete_bookmark(group, bookmark_name)
        success = bookmarks_handler.add_bookmark(group, new_name, bookmark_config)
    else:
        success = bookmarks_handler.update_bookmark(group, bookmark_name, bookmark_config)

    if not success:
        raise HTTPException(status_code=500, detail="Failed to update bookmark")

    return {"message": "Bookmark updated successfully"}

@router.delete("/{group}/{bookmark_name}")
async def delete_bookmark(group: str, bookmark_name: str):
    """Delete a bookmark"""
    success = bookmarks_handler.delete_bookmark(group, bookmark_name)

    if not success:
        raise HTTPException(status_code=404, detail="Bookmark or group not found")

    return {"message": "Bookmark deleted successfully"}

@router.post("/groups/{group}")
async def create_group(group: str):
    """Create a new bookmark group"""
    bookmarks = bookmarks_handler.load_bookmarks()
    groups = bookmarks_handler.parse_bookmarks(bookmarks)

    if group in groups:
        raise HTTPException(status_code=400, detail="Group already exists")

    # Add empty group
    groups[group] = []
    new_config = bookmarks_handler.build_bookmarks_config(groups)
    success = bookmarks_handler.save_bookmarks(new_config)

    if not success:
        raise HTTPException(status_code=500, detail="Failed to create group")

    return {"message": "Group created successfully"}

@router.put("/groups/{group}")
async def rename_group(group: str, new_name: str = Body(..., embed=True)):
    """Rename a bookmark group"""
    bookmarks = bookmarks_handler.load_bookmarks()
    groups = bookmarks_handler.parse_bookmarks(bookmarks)

    if group not in groups:
        raise HTTPException(status_code=404, detail="Group not found")

    if new_name in groups:
        raise HTTPException(status_code=400, detail="Group with new name already exists")

    # Rename the group
    groups[new_name] = groups.pop(group)
    new_config = bookmarks_handler.build_bookmarks_config(groups)
    success = bookmarks_handler.save_bookmarks(new_config)

    if not success:
        raise HTTPException(status_code=500, detail="Failed to rename group")

    return {"message": "Group renamed successfully"}

@router.delete("/groups/{group}")
async def delete_group(group: str):
    """Delete a bookmark group and all its bookmarks"""
    bookmarks = bookmarks_handler.load_bookmarks()
    groups = bookmarks_handler.parse_bookmarks(bookmarks)

    if group not in groups:
        raise HTTPException(status_code=404, detail="Group not found")

    del groups[group]
    new_config = bookmarks_handler.build_bookmarks_config(groups)
    success = bookmarks_handler.save_bookmarks(new_config)

    if not success:
        raise HTTPException(status_code=500, detail="Failed to delete group")

    return {"message": "Group deleted successfully"}

@router.get("/export")
async def export_bookmarks():
    """Export bookmarks configuration as YAML file"""
    yaml_content = bookmarks_handler.export_yaml()

    return Response(
        content=yaml_content,
        media_type="application/x-yaml",
        headers={
            "Content-Disposition": "attachment; filename=bookmarks.yaml"
        }
    )

@router.post("/import")
async def import_bookmarks(file: UploadFile = File(...)):
    """Import bookmarks from YAML file"""
    if not file.filename.endswith(('.yaml', '.yml')):
        raise HTTPException(status_code=400, detail="File must be a YAML file")

    try:
        contents = await file.read()
        yaml_content = contents.decode('utf-8')

        success = bookmarks_handler.import_yaml(yaml_content)

        if success:
            bookmarks = bookmarks_handler.load_bookmarks()
            groups = bookmarks_handler.parse_bookmarks(bookmarks)

            summary = {
                "message": "Bookmarks imported successfully",
                "groups": len(groups),
                "total_bookmarks": sum(len(bookmarks) for bookmarks in groups.values())
            }
            return summary
        else:
            raise HTTPException(status_code=500, detail="Failed to import bookmarks")

    except UnicodeDecodeError:
        raise HTTPException(status_code=400, detail="File encoding error")
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Import failed: {str(e)}")